### PR TITLE
[CUDA][cuBLASLt] torch.addmm - improve numerics of bias fusion with reduced precision SplitK algorithms

### PR DESCRIPTION
**Pull request by nikitaved**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #180875

After extensive statistical testing of cuBLASLt vs cuBLAS PyTorch dispatching paths, the culprit likely to cause numerical differences is found with high statistical confidence -- bias epilogue fusions with SplitK algorithms when run with reduced precision inputs. The problem is that bias is being added in the lower precision (cast to dtype of the result), and this seemingly innocuous detail can lead to quite dramatic numerical differences (not in the Lt's favor) in situations emulating proper distribution dynamics of hidden layers in modern NNs.

The solution is to back away from bias epilogue fusions in favor of bias fusions via a separate Matrix Descriptor. The Matrix Descriptor path preserves the compute dtype, and that is done in a higher precision dtype.

cc @ptrblck @msaroufim @eqy @jerryzh168 @tinglvv @nWEIdia @csarofeen

### CODE REVIEW

File: aten/src/ATen/cuda/CUDABlas.cpp [hunk 1]
```diff
@@ -1774,16 +1821,21 @@ bool gemm_and_bias(
...
         use_bias_descriptor = true;
-        if (std::holds_alternative<opmath_t>(beta)) {
-          beta = beta_val;
-        } else {
-          beta = static_cast<at::Half>(beta_val);
+        restore_beta_val();
+
+        // In the very, very unlikely event of not being able to re-use
+        // the selected algorithm, we need to re-query heuristic.
+        if (perform_algo_check() != CUBLAS_STATUS_SUCCESS) {
```
Comment (eqy) [inline aten/src/ATen/cuda/CUDABlas.cpp]: Should we use `C10_UNLIKELY` here?

**Comment by eqy** [comment]
@atalman would it be possible to get some signal on this internally? (after https://github.com/pytorch/pytorch/pull/179151 is landed)

**Comment by ngimel** [comment]
Let's back up a little - we were doing bias epilogue fusions for a long time even when cublas is used as a default backend (current state and state prior to 174954, so what changes when backend is set to cublasLt?

**Comment by nikitaved** [comment]
@ngimel , the Lt by default also affects the batched GEMM, this PR one only touches the addmm, not batched.

**Comment by ngimel** [comment]
I'm more confused. Pre-#174594 and #172340 we are doing bias fusion for addmm by default, this is our state for years, no one complains. Ignoring bmms (I believe most reported errors didn't concern bmms) what changed after #174594 and #172340?

**Comment by nikitaved** [comment]
Ignoring bmms... The difference is that "cublas" is the default backend in CIs, so it seems. Then https://github.com/pytorch/pytorch/pull/172340 disables bias fusions in the CIs (improves numerics), but then https://github.com/pytorch/pytorch/pull/174594 should bring it back (worsens numerics, relative to the previous PR). When the failures were reported, what is the reference point? If the reference point is https://github.com/pytorch/pytorch/pull/172340, then https://github.com/pytorch/pytorch/pull/174594 makes things worse, hence a revert. If THIS pr does not make things worse, we can analyze further -- it should improve numerics. Currently, it is either a) we have something else but addmm - baddbmm, maybe? or b) we caught the numerical change between the two aforementioned PRs or c) a combination of both.

**Comment by ngimel** [comment]
#172340 is NOT a reference point, we lived without #172340 for years and no one complained.

**Comment by nikitaved** [comment]
The failed test you pointed to in https://github.com/pytorch/pytorch/pull/174594#issuecomment-4027340890 I was able to fail on A100 (not on H100  -- my typical environment) with a switch to Lt as default -- the upstream status right at the merge of the problematic PR, but the tests were untouched for fair comparison. I logged two addmms the test run and the difference was huge due to different algorithms selected due to the workspace difference  -- none of these are bias-fused gemms, the Lt path is ignored inside addmm. Now, with the cherry-picked change https://github.com/pytorch/pytorch/pull/173002 "cublas" and "cublaslt" produce bitperfect results and the test passes. So, ignoring bmms, the status before and right after https://github.com/pytorch/pytorch/pull/174594 is that we used a smaller workspace due to the 1MB restriction for Lt vs 32MB for cuBLAS. Bringing workspace sizes to the common denominator fixed the issue. Batched gemms are likely not immune to this switch.

Note, however, the workspace unifying change in https://github.com/pytorch/pytorch/pull/173002 ignores FB_CODE, and that creates disparity between Lt and non-Lt default when it comes to workspace sizes, for the internal tests, unless unified workspace use is set explicitly. The pre-existing bias fusions will likely benefit from having 32MB vs 1MB workspace size as better algorithms are to be chosen.

This PR, however, is based on simulations run **with** and **without** unified workspace for typical shapes/distributions seen in modern NNs with an an additional controlled variable being whether addmm takes the Lt or the cuBLAS PyTorch path. In **either** of these cases bias fusion with SplitK produced significantly worse (despite normalizations NNs do) results compared to a double reference with p-values almost zero, so it was likely not random. We were able to confirm with the cuBLAS team that the epilogue bias fusion is done without the upcast when SplitK algorithm is selected. I cannot say whether better numerics would lead to better model performance due to gemm-related implicit biases during the training process  -- that could be tested internally -- ideally, with and without unified workspace. All in all, I agree that this PR addresses an orthogonal problem to what we had with Lt set to default, but it is still important to consider, as the running assumption is that bias fusions are done in higher precision -- not always the case, unfortunately.

**Comment by nikitaved** [comment]
@ngimel , the cuBLAS team told me that the bias fusion fix should be available in CTK 13.3.1. Shall we revive this patch?

**Comment by ngimel** [comment]
I guess when pytorch has 13.4 build?
